feat: Start implement base link handlers for new universal link-router system

smilingkylan opened the pull request:
Fixes: https://consensyssoftware.atlassian.net/jira/polaris/projects/MWMR/ideas/view/8845676?selectedIssue=MWMR-10

## **Description**

This PR implements the handler layer and integration for the Universal Router system, building on the core infrastructure from Phase 2 PR 1. It introduces a declarative, handler-based approach for processing deep links, replacing the need to modify core routing logic when adding new deep link types.

### What Changed?

Added handler implementation and integration layer to `app/core/DeeplinkManager/router/`:

1. **BaseHandler** (`handlers/BaseHandler.ts` - 103 lines)
   - Common utilities for all handlers (navigation, auth, error handling, analytics)
   - Parameter validation helpers
   - Success/error result factory methods

2. **NavigationHandler** (`handlers/NavigationHandler.ts` - 75 lines)
   - Handles simple navigation deep links (HOME, CREATE_ACCOUNT, REWARDS, PREDICT, PERPS)
   - Priority: 10 (standard)

3. **SwapHandler** (`handlers/SwapHandler.ts` - 67 lines)
   - Handles token swap deep links
   - Priority: 50 (high priority for core functionality)
   - Authentication-aware with default slippage settings

4. **SendHandler** (`handlers/SendHandler.ts` - 78 lines)
   - Handles transaction/send deep links
   - Delegates APPROVE actions to legacy system (requires complex transaction creation)
   - Parameter validation for required fields

5. **UniversalRouterIntegration** (`integration/UniversalRouterIntegration.ts` - 80 lines)
   - Integration wrapper for easy adoption into existing deeplink system
   - Feature flag checking (`MM_UNIVERSAL_ROUTER`)
   - Context creation from existing DeeplinkManager

6. **Updated UniversalRouter** (~20 lines added)
   - Registers all handlers in `initialize()` method
   - Imports and instantiates concrete handler classes

### Why?

Completes the Universal Router architecture started in Phase 2 PR 1 by:
- ✅ **Enabling handler-based routing**: New deep link types can be added without modifying core logic
- ✅ **Isolated testing**: Each handler can be tested independently
- ✅ **Clear separation**: Routing logic separated from handler implementation
- ✅ **Feature-flagged**: Safe gradual rollout with automatic fallback to legacy

### Trade-offs

- **No BrowserHandler yet**: Kept scope minimal. Browser/dApp links continue using legacy system for now.
- **APPROVE delegates to legacy**: Transaction creation is complex and tightly coupled to legacy system. Keeping it delegated reduces risk.
- **String route names**: Used string literals instead of Routes constants in some places to avoid circular dependencies.

## **Changelog**

CHANGELOG entry: null

_(Internal infrastructure - no end-user-facing behavior changes. Feature flag is OFF by default.)_

## **Related issues**

Refs: Deep Link Consolidation Epic (Phase 2 PR 2)

## **Manual testing steps**

Feature: Universal Router Handler System

  Scenario: Router remains inactive when feature flag is OFF
    Given the app is running with MM_UNIVERSAL_ROUTER flag disabled
    And handlers are registered in the UniversalRouter

    When a deep link is opened (e.g., metamask://home)
    Then the deep link should be handled by the legacy system
    And no errors should occur

  Scenario: Router processes links when feature flag is ON
    Given the app is running with MM_UNIVERSAL_ROUTER flag enabled
    And NavigationHandler is registered for HOME action

    When a deep link is opened (metamask://home)
    Then the UniversalRouter should route to NavigationHandler
    And the handler should navigate to home screen
    And analytics should track the handler success

  Scenario: Router falls back to legacy for unhandled actions
    Given the MM_UNIVERSAL_ROUTER flag is enabled
    And no handler is registered for a specific action

    When a deep link with unhandled action is opened
    Then the router should delegate to legacy system
    And no errors should occur

  Scenario: Authentication check blocks unauthenticated users
    Given the MM_UNIVERSAL_ROUTER flag is enabled
    And the wallet is locked

    When a deep link requiring authentication is opened (e.g., metamask://swap)
    Then the handler should detect authentication is required
    And should delegate to legacy system### Testing Commands

# Run all router tests
yarn jest app/core/DeeplinkManager/router/

# Run ESLint
yarn eslint app/core/DeeplinkManager/router/

# Run TypeScript check
yarn lint:tsc | grep "app/core/DeeplinkManager/router"## **Screenshots/Recordings**

N/A - Infrastructure change with no UI impact

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable (37 passing unit tests)
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---


<!-- CURSOR_SUMMARY -->
> [!NOTE]
> Adds handler-based routing (navigation, swap, send), registers them in `UniversalRouter`, and integrates behind a feature flag with comprehensive tests.
> 
> - **Router**
>   - `UniversalRouter.initialize()` now clears registry and registers `SwapHandler`, `SendHandler` (priority 50) and `NavigationHandler` (priority 10); throws on registration failure.
>   - Routing delegates to legacy when no handler matches or when a handler requests `fallbackToLegacy`.
> - **Handlers**
>   - Add `handlers/BaseHandler` with navigation/auth/analytics helpers and result/param utilities.
>   - Add `handlers/NavigationHandler` for actions like `HOME`, `CREATE_ACCOUNT`, `REWARDS`, `PREDICT`, `PERPS*`.
>   - Add `handlers/SwapHandler` and `handlers/SendHandler` with auth checks; `APPROVE` delegates to legacy.
>   - Export all via `handlers/index`.
> - **Integration**
>   - Add `integration/UniversalRouterIntegration` to gate usage via `MM_UNIVERSAL_ROUTER` and construct handler context; exposes `processWithNewRouter` and `getRouter`.
> - **Interfaces/Utils**
>   - Update `HandlerContext.instance` typing and docs; add optional lifecycle hooks; add `testUtils`.
> - **Tests**
>   - New unit tests for handlers and integration; update `UniversalRouter.test` (init failure case, priority order, legacy delegation).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7d8fe2d5767dc1ff25603a9b852540a270a2b0ec. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

smilingkylan committed:
Implement new link-handlers for universal link-router

smilingkylan committed:
Add more tests to boost coverage for new link-handlers

smilingkylan committed:
Remove unnecessary unit tests

smilingkylan committed:
Remove all tests for replacement

smilingkylan committed:
Re-add unit tests to link-handling

smilingkylan committed:
Fix unit test utility

smilingkylan committed:
Add a couple more new link-handler unit tests

smilingkylan committed:
Clean up new link handlers and related code

smilingkylan committed:
Add some more universal link router unit tests

smilingkylan committed:
Fix router initialization issue

smilingkylan committed:
Remove unit test

weitingsun commented on app/core/DeeplinkManager/router/handlers/BaseHandler.ts:
```diff
@@ -0,0 +1,111 @@
+    // Check if wallet is unlocked via Engine
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const Engine = (context.instance as any)?.context?.KeyringController;
```
This always returns undefined even if the wallet is unlocked so isAuthenticated is always false

weitingsun commented on app/core/DeeplinkManager/router/handlers/NavigationHandler.ts:
```diff
@@ -0,0 +1,83 @@
+          break;
+
+        case ACTIONS.PREDICT:
+          this.navigate(context, Routes.PREDICT.ROOT, link.params);
```
what about the market id logic in handlePredictUrl.ts?

smilingkylan replied:
Same

weitingsun commented on app/core/DeeplinkManager/router/handlers/NavigationHandler.ts:
```diff
@@ -0,0 +1,83 @@
+
+        case ACTIONS.PERPS:
+        case ACTIONS.PERPS_MARKETS:
+        case ACTIONS.PERPS_ASSET: {
```
same here with the logic in handlePerpsUrl.ts

weitingsun reviewed (commented):
@smilingkylan I have some questions, assuming these are something you'll address in follow up PRs? Maybe add TODOs so they are not forgotten.

smilingkylan committed:
Fix isAuthenticated mock for deeplink tests and clarify pending handler work

smilingkylan committed:
Fix Keyring controller reference

smilingkylan committed:
Fix deeplink handler unit tests (keyring controller)

smilingkylan committed:
Merge branch 'main' into kylan/feat/new-link-handlers

smilingkylan committed:
More deeplink manager unit test fixes

smilingkylan committed:
test: remove defensive edge case tests from UniversalRouterIntegration

Removed the 'shouldUseNewRouter - edge cases' describe block containing 3 tests that were testing defensive error handling for Redux state failures and feature flag selector errors. These tests were the least critical as they test unlikely error scenarios.

smilingkylan committed:
Remove optional chaining from new link router navigation call
